File: backend/internal/server/routes/payment.go
package routes

import (
	"github.com/Wei-Shaw/sub2api/internal/handler"
	"github.com/Wei-Shaw/sub2api/internal/handler/admin"
	"github.com/Wei-Shaw/sub2api/internal/server/middleware"
	"github.com/Wei-Shaw/sub2api/internal/service"

	"github.com/gin-gonic/gin"
)

// RegisterPaymentRoutes registers all payment-related routes:
// user-facing endpoints, webhook endpoints, and admin endpoints.
func RegisterPaymentRoutes(
	v1 *gin.RouterGroup,
	paymentHandler *handler.PaymentHandler,
	webhookHandler *handler.PaymentWebhookHandler,
	adminPaymentHandler *admin.PaymentHandler,
	jwtAuth middleware.JWTAuthMiddleware,
	adminAuth middleware.AdminAuthMiddleware,
	settingService *service.SettingService,
) {
	// --- User-facing payment endpoints (authenticated) ---
	authenticated := v1.Group("/payment")
	authenticated.Use(gin.HandlerFunc(jwtAuth))
	authenticated.Use(middleware.BackendModeUserGuard(settingService))
	{
		authenticated.GET("/config", paymentHandler.GetPaymentConfig)
		authenticated.GET("/checkout-info", paymentHandler.GetCheckoutInfo)
		authenticated.GET("/plans", paymentHandler.GetPlans)
		authenticated.GET("/channels", paymentHandler.GetChannels)
		authenticated.GET("/limits", paymentHandler.GetLimits)

		orders := authenticated.Group("/orders")
		{
			orders.POST("", paymentHandler.CreateOrder)
			orders.POST("/verify", paymentHandler.VerifyOrder)
			orders.GET("/my", paymentHandler.GetMyOrders)
			orders.GET("/:id", paymentHandler.GetOrder)
			orders.POST("/:id/cancel", paymentHandler.CancelOrder)
			orders.POST("/:id/refund-request", paymentHandler.RequestRefund)
			orders.GET("/refund-eligible-providers", paymentHandler.GetRefundEligibleProviders)
		}

		invoices := authenticated.Group("/invoices")
		{
			invoices.GET("/summary", paymentHandler.GetInvoiceSummary)
			invoices.GET("/requests", paymentHandler.ListInvoiceRequests)
			invoices.POST("/requests", paymentHandler.CreateInvoiceRequest)
			invoices.GET("/profiles", paymentHandler.ListInvoiceProfiles)
			invoices.POST("/profiles", paymentHandler.CreateInvoiceProfile)
			invoices.PUT("/profiles/:id", paymentHandler.UpdateInvoiceProfile)
			invoices.DELETE("/profiles/:id", paymentHandler.DeleteInvoiceProfile)
		}
	}

	sales := v1.Group("/sales")
	sales.Use(gin.HandlerFunc(jwtAuth))
	sales.Use(middleware.BackendModeUserGuard(settingService))
	{
		sales.GET("/dashboard", paymentHandler.GetSalesDashboard)
		sales.GET("/customers", paymentHandler.ListSalesCustomers)
		sales.GET("/orders", paymentHandler.GetSalesOrders)
		sales.GET("/customers/:id", paymentHandler.GetSalesCustomer)
		sales.GET("/customers/:id/orders", paymentHandler.GetSalesCustomerOrders)
	}

	// --- Public payment endpoints (no auth) ---
	// Signed resume-token recovery is the preferred public lookup path.
	// The legacy anonymous out_trade_no verify endpoint remains available as a
	// persisted-state compatibility path for staggered upgrades.
	public := v1.Group("/payment/public")
	{
		public.POST("/orders/verify", paymentHandler.VerifyOrderPublic)
		public.POST("/orders/resolve", paymentHandler.ResolveOrderPublicByResumeToken)
	}

	// --- Webhook endpoints (no auth) ---
	webhook := v1.Group("/payment/webhook")
	{
		// EasyPay sends GET callbacks with query params
		webhook.GET("/easypay", webhookHandler.EasyPayNotify)
		webhook.POST("/easypay", webhookHandler.EasyPayNotify)
		webhook.POST("/alipay", webhookHandler.AlipayNotify)
		webhook.POST("/wxpay", webhookHandler.WxpayNotify)
		webhook.POST("/stripe", webhookHandler.StripeWebhook)
		webhook.POST("/airwallex", webhookHandler.AirwallexWebhook)
	}

	// --- Admin payment endpoints (admin auth) ---
	adminGroup := v1.Group("/admin/payment")
	adminGroup.Use(gin.HandlerFunc(adminAuth))
	adminGroup.Use(middleware.AdminComplianceGuard(settingService))
	{
		// Dashboard
		adminGroup.GET("/dashboard", adminPaymentHandler.GetDashboard)

		// Config
		adminGroup.GET("/config", adminPaymentHandler.GetConfig)
		adminGroup.PUT("/config", adminPaymentHandler.UpdateConfig)
		adminGroup.GET("/invoice-config", adminPaymentHandler.GetInvoiceConfig)
		adminGroup.PUT("/invoice-config", adminPaymentHandler.UpdateInvoiceConfig)

		// Orders
		adminOrders := adminGroup.Group("/orders")
		{
			adminOrders.GET("", adminPaymentHandler.ListOrders)
			adminOrders.GET("/:id", adminPaymentHandler.GetOrderDetail)
			adminOrders.POST("/:id/cancel", adminPaymentHandler.CancelOrder)
			adminOrders.POST("/:id/retry", adminPaymentHandler.RetryFulfillment)
			adminOrders.POST("/:id/refund", adminPaymentHandler.ProcessRefund)
		}

		// Sales data visibility for admins
		adminSales := adminGroup.Group("/sales/:sales_id")
		{
			adminSales.GET("/dashboard", adminPaymentHandler.GetSalesDashboard)
			adminSales.GET("/customers", adminPaymentHandler.ListSalesCustomers)
			adminSales.GET("/orders", adminPaymentHandler.GetSalesOrders)
			adminSales.GET("/customers/:customer_id/orders", adminPaymentHandler.GetSalesCustomerOrders)
			adminSales.GET("/customers/:customer_id", adminPaymentHandler.GetSalesCustomer)
		}

		// Subscription Plans
		plans := adminGroup.Group("/plans")
		{
			plans.GET("", adminPaymentHandler.ListPlans)
			plans.POST("", adminPaymentHandler.CreatePlan)
			plans.PUT("/:id", adminPaymentHandler.UpdatePlan)
			plans.DELETE("/:id", adminPaymentHandler.DeletePlan)
		}

		// Provider Instances
		providers := adminGroup.Group("/providers")
		{
			providers.GET("", adminPaymentHandler.ListProviders)
			providers.POST("", adminPaymentHandler.CreateProvider)
			providers.PUT("/:id", adminPaymentHandler.UpdateProvider)
			providers.DELETE("/:id", adminPaymentHandler.DeleteProvider)
		}
	}
}

File: frontend/src/api/payment.ts
/**
 * User Payment API endpoints
 * Handles payment operations for regular users
 */

import { apiClient } from './client'
import type {
  PaymentConfig,
  SubscriptionPlan,
  PaymentChannel,
  MethodLimitsResponse,
  CheckoutInfoResponse,
  CreateOrderRequest,
  CreateOrderResult,
  PaymentOrder,
  InvoiceSummary,
  InvoiceProfile,
  InvoiceRequest,
  InvoiceProfilePayload,
  CreateInvoiceRequestPayload
} from '@/types/payment'
import type { BasePaginationResponse } from '@/types'

export const paymentAPI = {
  /** Get payment configuration (enabled types, limits, etc.) */
  getConfig() {
    return apiClient.get<PaymentConfig>('/payment/config')
  },

  /** Get available subscription plans */
  getPlans() {
    return apiClient.get<SubscriptionPlan[]>('/payment/plans')
  },

  /** Get available payment channels */
  getChannels() {
    return apiClient.get<PaymentChannel[]>('/payment/channels')
  },

  /** Get all checkout page data in a single call */
  getCheckoutInfo() {
    return apiClient.get<CheckoutInfoResponse>('/payment/checkout-info')
  },

  /** Get payment method limits and fee rates */
  getLimits() {
    return apiClient.get<MethodLimitsResponse>('/payment/limits')
  },

  /** Create a new payment order */
  createOrder(data: CreateOrderRequest) {
    return apiClient.post<CreateOrderResult>('/payment/orders', data)
  },

  /** Get current user's orders */
  getMyOrders(params?: { page?: number; page_size?: number; status?: string }) {
    return apiClient.get<BasePaginationResponse<PaymentOrder>>('/payment/orders/my', { params })
  },

  /** Get a specific order by ID */
  getOrder(id: number) {
    return apiClient.get<PaymentOrder>(`/payment/orders/${id}`)
  },

  /** Cancel a pending order */
  cancelOrder(id: number) {
    return apiClient.post(`/payment/orders/${id}/cancel`)
  },

  /** Verify order payment status with upstream provider */
  verifyOrder(outTradeNo: string) {
    return apiClient.post<PaymentOrder>('/payment/orders/verify', { out_trade_no: outTradeNo })
  },

  /** Legacy-compatible public order lookup by out_trade_no */
  verifyOrderPublic(outTradeNo: string) {
    return apiClient.post<PaymentOrder>('/payment/public/orders/verify', { out_trade_no: outTradeNo })
  },

  /** Resolve an order from a signed resume token without auth */
  resolveOrderPublicByResumeToken(resumeToken: string) {
    return apiClient.post<PaymentOrder>('/payment/public/orders/resolve', { resume_token: resumeToken })
  },

  /** Request a refund for a completed order */
  requestRefund(id: number, data: { reason: string }) {
    return apiClient.post(`/payment/orders/${id}/refund-request`, data)
  },

  /** Get provider instance IDs that allow user refund */
  getRefundEligibleProviders() {
    return apiClient.get<{ provider_instance_ids: string[] }>('/payment/orders/refund-eligible-providers')
  },

  /** Get current user's invoice quota summary */
  getInvoiceSummary() {
    return apiClient.get<InvoiceSummary>('/payment/invoices/summary')
  },

  /** Get current user's invoice application history */
  getInvoiceRequests(params?: { page?: number; page_size?: number; status?: string }) {
    return apiClient.get<BasePaginationResponse<InvoiceRequest>>('/payment/invoices/requests', { params })
  },

  /** Submit a new invoice application */
  createInvoiceRequest(data: CreateInvoiceRequestPayload) {
    return apiClient.post<InvoiceRequest>('/payment/invoices/requests', data)
  },

  /** Get current user's invoice titles */
  getInvoiceProfiles() {
    return apiClient.get<InvoiceProfile[]>('/payment/invoices/profiles')
  },

  /** Create an invoice title */
  createInvoiceProfile(data: InvoiceProfilePayload) {
    return apiClient.post<InvoiceProfile>('/payment/invoices/profiles', data)
  },

  /** Update an invoice title */
  updateInvoiceProfile(id: number, data: InvoiceProfilePayload) {
    return apiClient.put<InvoiceProfile>(`/payment/invoices/profiles/${id}`, data)
  },

  /** Delete an invoice title */
  deleteInvoiceProfile(id: number) {
    return apiClient.delete(`/payment/invoices/profiles/${id}`)
  }
}
